# PI_MLOPS.

 comenzando con este proyecto primero me dispongo a importar las librerias que voy a utilizar.

In [149]:
import pandas as pd
import ast

me doy cuenta que el archivo que me entregaron es  un archivo .json, entonces realizo una iteracion en cada una de las filas para meter la informacion de cada una de las filas en un a lista vacia que llamo 'rows' para luego convertir esa lista en un Data Frame de pandas que me servira para visulizar los datos, para mas adelante no tener problemas con leer el archivo .json tome la decisión de crear una archivo .csv que es exactamente igual al inicial que usare de aqui en adelante 

In [150]:
rows=[]
with open("Data/steam_games.json") as f:
    for line in f.readlines():
        rows.append(ast.literal_eval(line))
df_games = pd.DataFrame(rows)

In [151]:
df_games.to_csv('steam_games.csv', index=False)

In [152]:
df_games.head(1)

publisher                                         genres  \
0  Kotoshiro  [Action, Casual, Indie, Simulation, Strategy]   

              app_name                title  \
0  Lost Summoner Kitty  Lost Summoner Kitty   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   

                                            tags  discount_price  \
0  [Strategy, Action, Indie, Casual, Simulation]            4.49   

                                         reviews_url            specs price  \
0  http://steamcommunity.com/app/761140/reviews/?...  [Single-player]  4.99   

   early_access      id  developer sentiment metascore  
0         False  761140  Kotoshiro       NaN       NaN

In [153]:
df_games.isna().sum()

publisher          8052
genres             3283
app_name              2
title              2050
url                   0
release_date       2067
tags                163
discount_price    31910
reviews_url           2
specs               670
price              1377
early_access          0
id                    2
developer          3299
sentiment          7182
metascore         29458
dtype: int64

me doy cuenta que existen muchos valores faltantes en la gran mayoria de las columnas  

In [154]:
df_games[df_games["app_name"].isna()]

publisher           genres app_name title  \
74         NaN              NaN      NaN   NaN   
2580       NaN  [Action, Indie]      NaN   NaN   

                                              url release_date  \
74                 http://store.steampowered.com/          NaN   
2580  http://store.steampowered.com/app/317160/_/   2014-08-26   

                 tags  discount_price  \
74                NaN           14.99   
2580  [Action, Indie]             NaN   

                                            reviews_url  \
74                                                  NaN   
2580  http://steamcommunity.com/app/317160/reviews/?...   

                           specs  price  early_access      id developer  \
74                           NaN  19.99         False     NaN       NaN   
2580  [Single-player, Game demo]    NaN         False  317160       NaN   

     sentiment metascore  
74         NaN       NaN  
2580  Positive       NaN

In [155]:
df_games[df_games["id"].isna()]

publisher               genres  \
74                                                   NaN                  NaN   
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
74                                                NaN   
30961  Batman: Arkham City - Game of the Year Edition   

                                                title  \
74                                                NaN   
30961  Batman: Arkham City - Game of the Year Edition   

                                            url release_date  \
74               http://store.steampowered.com/          NaN   
30961  http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags  discount_price  \
74                                                   NaN           14.99   
30961  [Action, Open World, Batman, Adventure, Stealt...             NaN   

      reviews_url                                              specs  price  \
74            NaN                                                NaN  19.99   
30961         NaN  [Single-player, Steam Achievements, Steam Trad...  19.99   

       early_access   id                                   developer  \
74            False  NaN                                         NaN   
30961         False  NaN  Rocksteady Studios,Feral Interactive (Mac)   

                     sentiment metascore  
74                         NaN       NaN  
30961  Overwhelmingly Positive        91

luego de ver que la columna 'app_name' solamente tenia dos valores nulos decido buscarlos y con la url y buscando en google me doy cuenta que el registro 74 tienemuchos valores faltantes y no es relevante porque no tiene un nombre de juego, id ni una url relevante decido borrarlo y al registro 2580 luego de hacer el mismo procedimiento decido actualizar la columna 'app_name' y 'title'con el nombre del juego es 'DUET', tambien me doy cuenta que unicamente tengo dos valores faltantes en 'id' y ya borre uno de los faltantes entonces tambien decido buscar con el url y me doy cuenta que el id tambien me aparece en la url de steam  entonces procedo tambien a cambiarlo

In [156]:
# elimino la fila 74
df_games = df_games.drop(74)

In [157]:
df_games.loc[2580, 'app_name'] = "Duet" # Cambio el nombre en la columna 'app_name'
df_games.loc[2580, 'title'] = "Duet" # Cambio el nombre en la columna 'title'

In [158]:
df_games.loc[30961, 'id'] = 200260.0 # Cambio el Nan de  la columna 30961 en la columna 'id' por el valor 200260.0 porque me aparece en internet y esta en float

In [159]:
df_games.isna().sum()

publisher          8051
genres             3282
app_name              0
title              2048
url                   0
release_date       2066
tags                162
discount_price    31910
reviews_url           1
specs               669
price              1377
early_access          0
id                    0
developer          3298
sentiment          7181
metascore         29457
dtype: int64

In [160]:
df_games.head(3)

publisher                                             genres  \
0         Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1  Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2      Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   

                  app_name                    title  \
0      Lost Summoner Kitty      Lost Summoner Kitty   
1                Ironbound                Ironbound   
2  Real Pool 3D - Poolians  Real Pool 3D - Poolians   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1  http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2  http://store.steampowered.com/app/670290/Real_...   2017-07-24   

                                                tags  discount_price  \
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2  [Free to Play, Simulation, Sports, Casual, Ind...             NaN   

                                         reviews_url  \
0  http://steamcommunity.com/app/761140/reviews/?...   
1  http://steamcommunity.com/app/643980/reviews/?...   
2  http://steamcommunity.com/app/670290/reviews/?...   

                                               specs         price  \
0                                    [Single-player]          4.99   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
2  [Single-player, Multi-player, Online Multi-Pla...  Free to Play   

   early_access      id         developer        sentiment metascore  
0         False  761140         Kotoshiro              NaN       NaN  
1         False  643980  Secret Level SRL  Mostly Positive       NaN  
2         False  670290      Poolians.com  Mostly Positive       NaN

In [161]:
df_games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32134 entries, 0 to 32134
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   publisher       24083 non-null  object 
 1   genres          28852 non-null  object 
 2   app_name        32134 non-null  object 
 3   title           30086 non-null  object 
 4   url             32134 non-null  object 
 5   release_date    30068 non-null  object 
 6   tags            31972 non-null  object 
 7   discount_price  224 non-null    float64
 8   reviews_url     32133 non-null  object 
 9   specs           31465 non-null  object 
 10  price           30757 non-null  object 
 11  early_access    32134 non-null  bool   
 12  id              32134 non-null  object 
 13  developer       28836 non-null  object 
 14  sentiment       24953 non-null  object 
 15  metascore       2677 non-null   object 
dtypes: bool(1), float64(1), object(14)
memory usage: 5.0+ MB


In [162]:
df_games["id"] = df_games["id"].astype(int)

In [163]:
df_games["id"].info()

<class 'pandas.core.series.Series'>
Int64Index: 32134 entries, 0 to 32134
Series name: id
Non-Null Count  Dtype
--------------  -----
32134 non-null  int32
dtypes: int32(1)
memory usage: 1.4 MB


In [164]:
df_games['release_date'] = pd.to_datetime(games['release_date'], errors='coerce')

luego de ver la informacion general del dataframe( .info()) cambio el tipo de datos de la columna 'id' de float a int, y la columna 'release_date' a formato fecha, luego creo la columna nueva de 'release_year' donde almacenare el año 

In [165]:
df_games["release_year"] = df_games["release_date"].dt.year

In [269]:
df_games['release_year'].fillna(9999, inplace=True)

In [286]:
df_games['release_year'] = df_games["release_year"].astype(int)

luego de hacer las pocas transformaciones necesarias voy a comenzar a crear un archivo '.csv' para cada una de las funciones que guardare en la carpeta 'Data' creo un dataframe donde podre consultar mas adelante las funciones utilizando unicamente las columnas que necesito, sin alterar el archivo original, en cada uno de estos datraframes nuevos voy a a eliminar las filas que creo pertinentes para ofrecer la mejor calidad en  las consultas que seran consumidas por la API  y teniendo simpre presentes los requerimientos del proyecto.

## Genero:

voy crear un nuevo data frame por cada una de las funciones para luego cargarla como un csv, y voy a revisar los valores nulos y con relacion a la cantidad de registros deciso borrarlos, creo una funcion que mas tarde la pasare para pode correr mi API 

In [287]:
genero = df_games[['release_year','genres']]

In [288]:
genero.isna().sum()

release_year       0
genres          3282
dtype: int64

In [291]:
genero.dropna(subset=['genres'], inplace=True)

C:\Users\mateo\AppData\Local\Temp\ipykernel_15668\1812928130.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  genero.dropna(subset=['genres'], inplace=True)


In [292]:
genero.isna().sum()

release_year    0
genres          0
dtype: int64

In [293]:
genero.head()

release_year                                             genres
0          2018      [Action, Casual, Indie, Simulation, Strategy]
1          2018               [Free to Play, Indie, RPG, Strategy]
2          2017  [Casual, Free to Play, Indie, Simulation, Sports]
3          2017                        [Action, Adventure, Casual]
5          2018                    [Action, Adventure, Simulation]

In [294]:
genero.to_csv('Data/generos.csv', index=False)

In [487]:
df_genero= pd.read_csv('Data/generos.csv')

In [488]:
df_genero.head()

release_year                                             genres
0          2018  ['Action', 'Casual', 'Indie', 'Simulation', 'S...
1          2018       ['Free to Play', 'Indie', 'RPG', 'Strategy']
2          2017  ['Casual', 'Free to Play', 'Indie', 'Simulatio...
3          2017                  ['Action', 'Adventure', 'Casual']
4          2018              ['Action', 'Adventure', 'Simulation']

voy a aplicar de la libreria ast a la columna generos, ya que luego de crear un csv no se crean una lista sino cadenas y lo que necesito para poder utilizarlo son listas que mas adelante voy a realizar la operacion 'explode()'  

In [489]:
df_genero['genres'] = df_genero['genres'].apply(ast.literal_eval)

In [490]:
df_genero.head()

release_year                                             genres
0          2018      [Action, Casual, Indie, Simulation, Strategy]
1          2018               [Free to Play, Indie, RPG, Strategy]
2          2017  [Casual, Free to Play, Indie, Simulation, Sports]
3          2017                        [Action, Adventure, Casual]
4          2018                    [Action, Adventure, Simulation]

In [496]:
def funcion_generos(year):
    if year == 9999:
        print ("NO SE TIENE UN REGISTRO PRECISO DE LA LISTA DE GENEROS  PARA EL AÑO INGRESADO, ES UNA RECOPILACION DE LOS POSIBLES GENEROS ")
    df_anio = df_genero[df_genero['release_year'] == year]
    if df_anio.empty:
        return {"VALUE_ERROR": "NO EXISTEN DATOS DISPONIBLES PARA EL AÑO INGRESADO"}
    
    count_genres = df_anio['genres'].explode().value_counts().to_dict()

    top_genres = dict(sorted(count_genres.items(), key=lambda item: item[1], reverse=True)[:5])
    
    return top_genres

In [497]:
funcion_generos(2016)

{'Indie': 4105,
 'Action': 2543,
 'Casual': 2141,
 'Adventure': 2041,
 'Simulation': 1666}

creo la funcion de generos que pruebo para luego utilizarlo en el archivo'main.py' donde estara el codigo de mi API 

## Juegos:

voy a crear un dataframe unicamente con las columnas mas relevantes para esta funcion voy a utilizar las columnas 'app_name' y 'release_year'

In [301]:
juegos = df_games[['release_year','app_name']]

In [302]:
juegos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32134 entries, 0 to 32134
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   release_year  32134 non-null  int32 
 1   app_name      32134 non-null  object
dtypes: int32(1), object(1)
memory usage: 1.6+ MB


In [303]:
juegos.to_csv('Data/juegos.csv', index=False)

In [495]:
df_juegos = pd.read_csv('Data/juegos.csv')

voy a crear la funcion de juegos que me dira todos los juegos lanzados en un año ingresado 

In [499]:
def funcion_juegos(year):
    if year == 9999:
        print ("NO SE TIENE UN REGISTRO PRECISO DE LA LISTA DE JUEGOS  PARA EL AÑO INGRESADO, ES UNA RECOPILACION DE LOS POSIBLES JUEGOS ")
    df_anio = df_juegos[df_juegos['release_year'] == year]
    if df_anio.empty:
        return {"VALUE_ERROR": "NO EXISTEN DATOS DISPONIBLES PARA EL AÑO INGRESADO"}

    df_anio['app_name'] = df_anio['app_name'].astype(str)

    juegos_lanzados = df_anio['app_name'].astype(str).tolist()
    cantidad_juegos = len(juegos_lanzados)

    return {"Año": year, "Cantidad de Juegos": cantidad_juegos, "Juegos Lanzados": juegos_lanzados}
      

In [502]:
funcion_juegos(2000)

C:\Users\mateo\AppData\Local\Temp\ipykernel_15668\250903329.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_anio['app_name'] = df_anio['app_name'].astype(str)


{'Año': 2000,
 'Cantidad de Juegos': 40,
 'Juegos Lanzados': ['The Longest Journey',
  'Unreal Tournament: Game of the Year Edition',
  'Sacrifice',
  'X: Tension',
  'Fairy Tale About Father Frost, Ivan and Nastya',
  'Pajama Sam 3: You Are What You Eat From Your Head To Your Feet',
  'Putt-Putt® Joins the Circus',
  'Gunship!',
  'Toilet Tycoon',
  'Two Family House',
  'Ghost Dog: The Way of the Samurai',
  'Call to Power II',
  'Vampire: The Masquerade - Redemption',
  'Nancy Drew®: Message in a Haunted Mansion',
  'Beachhead 2000',
  'Hidden & Dangerous: Action Pack',
  'Invictus: In the Shadow of Olympus',
  'Leprechaun: In the Hood',
  'American Psycho',
  'Hogs of War',
  'Ground Control Anthology',
  'SHOGUN: Total War™ - Collection',
  'Puzzle Station 15th Anniversary Retro Release',
  'Atlantis 2: Beyond Atlantis',
  'Carmageddon TDR 2000',
  'B-17 Flying Fortress: The Mighty 8th',
  'Spy Fox In: Hold the Mustard',
  'Soulbringer',
  'Earth 2150 Trilogy',
  'Daikatana',
  'T

La funcion que decidi crear en el resultado final me da un diccionario del año que yo ingreso y le agregue una parte donde me dice cuantos juegos se lanzaron en ese año y la lista completa de todos los juegos lanzados en el año ingresado

## Specs:

sigo con la misma dinamica creando un dataframe unicamente con las columna necesarias: 'release_year' y 'specs' y luego lo convierto en un archiva .csv para finalmente crear la funcion.

In [314]:
especificaciones = df_games[['release_year','specs']]

In [315]:
especificaciones.isna().sum()

release_year      0
specs           669
dtype: int64

voy a borrar las filas que tengan los datos Nan de la columna 'specs' ya que relativamente no son muchos en comparacion con la cantidad de registros   

In [316]:
especificaciones.shape

(32134, 2)

In [318]:
especificaciones.dropna(subset=['specs'], inplace=True)

C:\Users\mateo\AppData\Local\Temp\ipykernel_15668\1715576812.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  especificaciones.dropna(subset=['specs'], inplace=True)


In [323]:
especificaciones.to_csv('Data/specs.csv', index=False)

In [324]:
df_specs = pd.read_csv('Data/specs.csv')

In [326]:
df_specs['specs'] = df_specs['specs'].apply(ast.literal_eval)

luego de crear el archivo .csv lo cargo para crear la funcion que mas tarde utiizare en mi archivo 'main.py'

In [503]:
def funcion_specs(year):
    if year == 9999:
        print ("NO SE TIENE UN REGISTRO PRECISO DE LA LISTA DE ESPECIFICACIONES  PARA EL AÑO INGRESADO, ES UNA RECOPILACION DE LAS POSIBLES ESPECIFICACIONES ")
    df_anio = df_specs[df_specs['release_year'] == year]
    if df_anio.empty:
        return {"VALUE_ERROR": "NO EXISTEN DATOS DISPONIBLES PARA EL AÑO INGRESADO"}

    count_specs = df_anio['specs'].explode().value_counts().to_dict()
    top_specs = dict(sorted(count_specs.items(), key=lambda item: item[1], reverse=True)[:5])

    return {'Año': year, 'Especificaciones': top_specs}

In [508]:
funcion_specs(2015)

{'Año': 2015,
 'Especificaciones': {'Single-player': 4354,
  'Steam Achievements': 2870,
  'Steam Trading Cards': 2521,
  'Downloadable Content': 2156,
  'Steam Cloud': 1888}}

Como en la primera funcion de generos decidi reutilizar el codigo en donde el resultado es el año que se ingresa y un diccionario de las especificaciones del mismo año con un conteo mas detallado de cada una de ellas 

## Early_Acces:

In [336]:
df_games.head(2)

publisher                                         genres  \
0         Kotoshiro  [Action, Casual, Indie, Simulation, Strategy]   
1  Making Fun, Inc.           [Free to Play, Indie, RPG, Strategy]   

              app_name                title  \
0  Lost Summoner Kitty  Lost Summoner Kitty   
1            Ironbound            Ironbound   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1  http://store.steampowered.com/app/643980/Ironb...   2018-01-04   

                                                tags  discount_price  \
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   

                                         reviews_url  \
0  http://steamcommunity.com/app/761140/reviews/?...   
1  http://steamcommunity.com/app/643980/reviews/?...   

                                               specs         price  \
0                                    [Single-player]          4.99   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   

   early_access      id         developer        sentiment metascore  \
0         False  761140         Kotoshiro              NaN       NaN   
1         False  643980  Secret Level SRL  Mostly Positive       NaN   

   release_year  
0          2018  
1          2018

In [337]:
earlyacces = df_games[['release_year','early_access']]

In [339]:
earlyacces.isna().sum()

release_year    0
early_access    0
dtype: int64

In [349]:
earlyacces.to_csv('Data/earlyaccess.csv', index=False)

In [350]:
df_earlyacces = pd.read_csv('Data/earlyaccess.csv')

In [362]:
df_earlyacces.head()

release_year  early_access
0          2018         False
1          2018         False
2          2017         False
3          2017         False
4          9999         False

In [512]:
def funcion_earlyacces(year):
   if year == 9999:
      print ("NO SE TIENE UN REGISTRO PRECISO DE LA LISTA DE ACCESOS TEMPRANOS  PARA EL AÑO INGRESADO, ES UNA RECOPILACION DE LOS POSIBLES ACCESOS TEMPRANOS ")
   filtro = (df_earlyacces['release_year'] == year) & (df_earlyacces['early_access'] == True)
   df_year_earlyacces = df_earlyacces[filtro]

   if df_year_earlyacces.empty:
      return{"VALUE_ERROR": "NO EXISTEN DATOS DISPONIBLES PARA EL AÑO INGRESADO, INTENTE CON OTRA FECHA"}

   count_earlyacces = df_year_earlyacces.shape[0]

   return {"Año": year, "Cantidad de Juegos lanzados con early access": count_earlyacces}

In [513]:
funcion_earlyacces(2014)

{'Año': 2014, 'Cantidad de Juegos lanzados con early access': 78}

## sentiment:

In [388]:
sentiments = df_games[['release_year', 'sentiment']]

In [390]:
sentiments.isna().sum()

release_year       0
sentiment       7181
dtype: int64

In [397]:
sentiments["sentiment"]

0                    NaN
1        Mostly Positive
2        Mostly Positive
3                    NaN
4                    NaN
              ...       
32130                NaN
32131                NaN
32132                NaN
32133     1 user reviews
32134           Positive
Name: sentiment, Length: 32134, dtype: object

In [399]:
sentimientos_unicos = sentiments["sentiment"].unique()
sentimientos_unicos

array([nan, 'Mostly Positive', 'Mixed', '1 user reviews',
       '3 user reviews', '8 user reviews', 'Very Positive',
       'Overwhelmingly Positive', '6 user reviews', '5 user reviews',
       '2 user reviews', 'Very Negative', 'Positive', 'Mostly Negative',
       '9 user reviews', 'Negative', '4 user reviews', '7 user reviews',
       'Overwhelmingly Negative'], dtype=object)

In [407]:
estandar_sentimientos = ['Overwhelmingly Positive','Very Positive','Mostly Positive','Positive','Mixed','Negative','Mostly Negative','Very Negative','Overwhelmingly Negative']

luego de ver todos los valores que se pueden encontrar en la columna de 'sentiment' decido crear una lista con los sentimientos que realmente me interesan y pienso que son mas relevantes, de esta manera va a ser mas facil crear el dataframe para poder consultarlo   

In [408]:
 sentiments = sentiments[sentiments['sentiment'].isin(estandar_sentimientos)]

In [412]:
sentiments.shape[0]

(15260, 2)

In [413]:
32134-15260

16874

luego de tener estandarizado este dataframe y haber eliminado 16874 filas que no eran relevantes para saber el 'sentiment' ya puedo crear el archivo .csv que utilizare para mi funcion de la Api de Sentiment

In [415]:
sentiments.to_csv('Data/sentiments.csv', index=False)

In [416]:
df_sentiment = pd.read_csv('Data/sentiments.csv')

In [514]:
def funcion_sentiment(year):
    if year == 9999:
        print ("NO SE TIENE UN REGISTRO PRECISO DE LA LISTA DE SENTIMIENTOS  PARA EL AÑO INGRESADO, ES UNA RECOPILACION DE LOS POSIBLES SENTIMIENTOS ")
    df_anio = df_sentiment[df_sentiment['release_year'] == year] 

    if df_anio.empty:
        return {"VALUE_ERROR": "NO EXISTEN DATOS DISPONIBLES PARA EL AÑO INGRESADO, INTENTE CON OTRA FECHA"}

    conteo_sentimientos = df_anio['sentiment'].value_counts().to_dict()

    return {'Año': year, 'Análisis de sentimiento': conteo_sentimientos}

In [516]:
funcion_sentiment(2012)

{'Año': 2012,
 'Análisis de sentimiento': {'Very Positive': 191,
  'Mixed': 148,
  'Positive': 111,
  'Mostly Positive': 101,
  'Mostly Negative': 27,
  'Overwhelmingly Positive': 20,
  'Negative': 4,
  'Very Negative': 2}}

cumpliendo con los requerimientos que me piden en el proyecto la respuesta que devuelvo de cada funcion es el diccionario con el anális de los diferentes sentimientos determinado por el año ingresado  

## Metascore:

In [432]:
df_games.head(2)

publisher                                         genres  \
0         Kotoshiro  [Action, Casual, Indie, Simulation, Strategy]   
1  Making Fun, Inc.           [Free to Play, Indie, RPG, Strategy]   

              app_name                title  \
0  Lost Summoner Kitty  Lost Summoner Kitty   
1            Ironbound            Ironbound   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1  http://store.steampowered.com/app/643980/Ironb...   2018-01-04   

                                                tags  discount_price  \
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   

                                         reviews_url  \
0  http://steamcommunity.com/app/761140/reviews/?...   
1  http://steamcommunity.com/app/643980/reviews/?...   

                                               specs         price  \
0                                    [Single-player]          4.99   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   

   early_access      id         developer        sentiment metascore  \
0         False  761140         Kotoshiro              NaN       NaN   
1         False  643980  Secret Level SRL  Mostly Positive       NaN   

   release_year  
0          2018  
1          2018

In [433]:
metascore = df_games[['release_year','app_name','metascore']]

In [439]:
metascore.isna().sum()

release_year        0
app_name            0
metascore       29457
dtype: int64

In [440]:
metascore

release_year                  app_name metascore
0              2018       Lost Summoner Kitty       NaN
1              2018                 Ironbound       NaN
2              2017   Real Pool 3D - Poolians       NaN
3              2017                   弹炸人2222       NaN
4              9999             Log Challenge       NaN
...             ...                       ...       ...
32130          2018            Colony On Mars       NaN
32131          2018  LOGistICAL: South Africa       NaN
32132          2018             Russian Roads       NaN
32133          2017       EXIT 2 - Directions       NaN
32134          9999               Maze Run VR       NaN

[32134 rows x 3 columns]

In [442]:
metascore['metascore'].notna().sum()

2677

luego de ver cuantos valores Nan tengo en el Dataset y realmente los valores que tienen alguna informacion es muy bajo puedo crear una csv con todos los valores o podria crear una data set unicamente con los valores que tengan algun tipo de valor valido , pero como no conosco realmente los valore voy a dejarlo de esta manera para crear el dataset y al monmento de hacer la funcion relizo los respectivos filtros pra no perder datos que podrian ser relevantes 

In [444]:
metascore.to_csv('Data/metascore.csv', index=False)

In [445]:
df_metascore = pd.read_csv('Data/metascore.csv')

In [448]:
df_metascore

release_year                  app_name  metascore
0              2018       Lost Summoner Kitty        NaN
1              2018                 Ironbound        NaN
2              2017   Real Pool 3D - Poolians        NaN
3              2017                   弹炸人2222        NaN
4              9999             Log Challenge        NaN
...             ...                       ...        ...
32129          2018            Colony On Mars        NaN
32130          2018  LOGistICAL: South Africa        NaN
32131          2018             Russian Roads        NaN
32132          2017       EXIT 2 - Directions        NaN
32133          9999               Maze Run VR        NaN

[32134 rows x 3 columns]

In [461]:
historico_mejor_metascore = df_metascore.nlargest(10, 'metascore')
historico_mejor_metascore

release_year                                           app_name  \
28             1998                                          Half-Life   
95             2007                                          BioShock™   
26954          2015                                 Grand Theft Auto V   
32015          2004                                        Half-Life 2   
631            2011                                           Portal 2   
92             2007                                              QUAKE   
252            2007  The Elder Scrolls IV: Oblivion® Game of the Ye...   
254            2009  The Elder Scrolls IV: Oblivion® Game of the Ye...   
561            2003                        Grand Theft Auto: Vice City   
4629           2015          Divinity: Original Sin - Enhanced Edition   

       metascore  
28          96.0  
95          96.0  
26954       96.0  
32015       96.0  
631         95.0  
92          94.0  
252         94.0  
254         94.0  
561         94.0  
4629        94.0

In [483]:
def funcion_metascore(year):
    if year == 9999:
        print ("NO SE TIENE UN REGISTRO CLARO DE LA LISTA DE JUEGOS ,NI SU AÑO DE LANZAMIENTO ")

    df_anio = df_metascore[df_metascore['release_year'] == year ]

    if df_anio['metascore'].notna().any():
        top_juegos = df_anio.nlargest(5, 'metascore')[['app_name', 'metascore']].to_dict('registros_records')

        return {'Año': year,'registros_records': top_juegos}
    else:
        return {"VALUE_ERROR": "NO EXISTEN REGISTROS DISPONIBLES PARA EL AÑO INGRESADO, INTENTE CON OTRA FECHA"}
    


In [486]:
funcion_metascore(9999)

NO SE TIENE UN REGISTRO CLARO DE LA LISTA DE JUEGOS ,NI SU AÑO DE LANZAMIENTO 


C:\Users\mateo\AppData\Local\Temp\ipykernel_15668\1640345666.py:8: FutureWarning: Using short name for 'orient' is deprecated. Only the options: ('dict', list, 'series', 'split', 'records', 'index') will be used in a future version. Use one of the above to silence this warning.
  top_juegos = df_anio.nlargest(5, 'metascore')[['app_name', 'metascore']].to_dict('registros_records')


{'Año': 9999,
 'registros_records': [{'app_name': 'Dota 2', 'metascore': 90.0},
  {'app_name': 'Rez Infinite', 'metascore': 89.0},
  {'app_name': 'Raw Data', 'metascore': 87.0},
  {'app_name': 'The Gallery - Episode 2: Heart of the Emberstone',
   'metascore': 86.0},
  {'app_name': 'Rise of the Tomb Raider™', 'metascore': 86.0}]}

en esta funcion en particular cuando me arroja la lista de los juegos y su diferente metascore quise tratar los valores de la columna 'release_year' de una manera diferente ya que durante de todo el ejercicio los valpors de la columna 'release_year' con un valor igual a '9999' fue una manera que inialmente decidi tratar los valores Nan pero ahora al momento de ingresar esta fecha en la funcion me sale un mensaje informativo de alerta que me dice que el metascore de esa fecha no necesariamente coicide con algun año en especifico  